swarm/storage/pyramid.go

```diff
@@ -73,128 +130,496 @@ func NewPyramidChunker(params *ChunkerParams) (self *PyramidChunker) {
-		results.Levels[i] = make(map[int64]*Node)
+	return rootKey, nil
+
+}
+
+func (self *PyramidChunker) Append(key Key, data io.Reader, chunkC chan *Chunk, storageWG, processorWG *sync.WaitGroup) (Key, error) {
+
+	quitC := make(chan bool)
+	rootKey := make([]byte, self.hashSize)
+	chunkLevel := make([][]*TreeEntry, self.branches)
+
+	// Load the right most unfinished tree chunks in every level
+	self.loadTree(chunkLevel, key, chunkC, quitC)
+
+	jobC := make(chan *chunkJob, 2*ChunkProcessors)
+	wg := &sync.WaitGroup{}
+	errC := make(chan error)
+
+	wg.Add(1)
+	go self.prepareChunks(true, chunkLevel, data, rootKey, quitC, wg, jobC, processorWG, chunkC, errC, storageWG)
+
+	// closes internal error channel if all subprocesses in the workgroup finished
+	go func() {
+
+		// waiting for all chunks to finish
+		wg.Wait()
+
+		// if storage waitgroup is non-nil, we wait for storage to finish too
+		if storageWG != nil {
+			storageWG.Wait()
+		}
+		close(errC)
+	}()
+
+	defer close(quitC)
+
+	select {
+	case err := <-errC:
+		if err != nil {
+			return nil, err
+		}
+	case <-time.NewTimer(splitTimeout).C:
 	}
-	// Create a pool of workers to crunch through the file
-	tasks := make(chan *Task, 2*processors)
-	pend := new(sync.WaitGroup)
-	abortC := make(chan bool)
-	for i := 0; i < processors; i++ {
-		pend.Add(1)
-		go self.processor(pend, swg, tasks, chunkC, &results)
+	return rootKey, nil
+
+}
+
+func (self *PyramidChunker) processor(id int64, jobC chan *chunkJob, chunkC chan *Chunk, errC chan error, quitC chan bool, swg, wwg *sync.WaitGroup) {
+
+	defer atomic.AddInt64(&self.workerCount, -1)
+
+	hasher := self.hashFunc()
+	if wwg != nil {
+		defer wwg.Done()
 	}
-	// Feed the chunks into the task pool
-	read := 0
-	for index := 0; ; index++ {
-		buffer := make([]byte, self.chunkSize+8)
-		n, err := data.Read(buffer[8:])
-		read += n
-		last := int64(read) == size || err == io.ErrUnexpectedEOF || err == io.EOF
-		if err != nil && !last {
-			close(abortC)
-			break
-		}
-		binary.LittleEndian.PutUint64(buffer[:8], uint64(n))
-		pend.Add(1)
+	for {
 		select {
-		case tasks <- &Task{Index: int64(index), Size: uint64(n), Data: buffer[:n+8], Last: last}:
-		case <-abortC:
-			return nil, err
+
+		case job, ok := <-jobC:
+			if !ok {
+				return
+			}
+			self.processChunk(id, hasher, job, chunkC, swg)
+		case <-quitC:
+			return
 		}
-		if last {
-			break
+	}
+}
+
+func (self *PyramidChunker) processChunk(id int64, hasher SwarmHash, job *chunkJob, chunkC chan *Chunk, swg *sync.WaitGroup) {
+
+	hasher.ResetWithLength(job.chunk[:8]) // 8 bytes of length
+	hasher.Write(job.chunk[8:])           // minus 8 []byte length
+	h := hasher.Sum(nil)
+
+	newChunk := &Chunk{
+		Key:   h,
+		SData: job.chunk,
+		Size:  job.size,
+		wg:    swg,
+	}
+
+	// report hash of this chunk one level up (keys corresponds to the proper subslice of the parent chunk)
+	copy(job.key, h)
+
+	// send off new chunk to storage
+	if chunkC != nil {
+		if swg != nil {
+			swg.Add(1)
```

Comment:
If we are sending chunks to a channel here, why do we also have to increment a "storage" wait group? Shouldn't it be the receiver of the channel that is actually going to store the chunk who should also manage this wait group?

Reply:
- The receiver function "storeWorker" in dpa uses only the channel to store the chunk.
- The swg in the above snippet is the chunk's own wg, which signals that the chunk is stored. This is used in testing to simulate a put.

The actual use of this is while storing in memstore, it is incremented and after it is stored in DB, it is decremented indicating that the operation is done. per chunk vise.